core/store/src/flat_state.rs

```diff
@@ -1014,6 +1037,19 @@ impl FlatStorageState {
         for block in blocks.into_iter().rev() {
             let mut store_update = StoreUpdate::new(guard.store.storage.clone());
             let delta = guard.get_delta(&block)?.as_ref().clone();
+            for (key, value) in delta.0.iter() {
+                // Note that capacity can be zero, and we need to treat the case when no value can be pushed/popped.
+                if guard.value_ref_cache.len() == guard.value_ref_cache.cap() {
+                    if let Some((key, _)) = guard.value_ref_cache.pop_lru() {
+                        guard.metrics.value_ref_cache_total_key_size.sub(key.len() as i64);
+                    }
+                }
+                if guard.value_ref_cache.len() < guard.value_ref_cache.cap() {
+                    guard.metrics.value_ref_cache_total_key_size.add(key.len() as i64);
+                    guard.value_ref_cache.put(key.clone(), value.clone());
+                }
```

Comment:
You may be able to simplify this code with the lru cache `push` method - it checks the capacity by itself and returns a popped item if any. Not sure how it would handle 0 capacity though. 

https://docs.rs/lru/latest/lru/struct.LruCache.html#method.push